packages/ckeditor5-engine/src/view/matcher.js

```diff
@@ -242,71 +244,183 @@ function matchName( pattern, name ) {
+			if (
+				isAttributeKeyMatched( patternKey, attributeKey ) &&
+				isAttributeValueMatched( patternValue, attributeValue )
+			) {
+				match.push( attributeKey );
+			}
+		} );
+	} );
+
+	// Return null when there was no matches or we didn't match all patterns.
+	if ( !match.length || match.length != patterns.length ) {
+		return null;
+	}
+
 	return match;
 }
 
+// Bring all the possible pattern forms to an array of objects with `key` and `value` property.
+// For example:
+//
+//	{
+//		src: /https:.*/
+//	}
+//
+// to:
+//
+//	{
+//		key: 'src',
+//		value: /https:.*/
+//	}
+// @param {Object|Array} patterns
+// @returns {Array|null} Returns an array of objects or null if provided patterns were not in an expected form.
+function normalizePatterns( patterns ) {
+	if ( Array.isArray( patterns ) ) {
+		return patterns.map(
+			pattern => {
+				// eslint-disable-next-line dot-notation
+				if ( isPlainObject( pattern ) && pattern[ 'key' ] && pattern[ 'value' ] ) {
+					return pattern;
+				}
+
+				// Assume the pattern is either String or RegExp.
+				return { key: pattern, value: true };
+			}
+		);
+	} else if ( isPlainObject( patterns ) ) {
+		return Object.entries( patterns ).map(
+			( [ patternKey, patternValue ] ) => ( { key: patternKey, value: patternValue } )
+		);
+	}
+
+	return null;
+}
+
+// @param {String|RegExp} patternKey A pattern representing attribute key we want to match.
+// @param {String} attributeKey An actual attribute key (e.g. `'src'`, `'background-color'`, `'ck-widget'`) we're testing against pattern.
+// @returns {Boolean}
+function isAttributeKeyMatched( patternKey, attributeKey ) {
+	return patternKey === attributeKey ||
+		( patternKey instanceof RegExp && patternKey.test( attributeKey ) );
+}
+
+// @param {String|RegExp} patternValue A pattern representing attribute value we want to match.
+// @param {String} attributeValue An actual attribute value (e.g. `'http://example.com/'`, `'red'`) we're testing against pattern.
+// @returns {Boolean}
+function isAttributeValueMatched( patternValue, attributeValue ) {
+	return patternValue === true ||
+		patternValue === attributeValue ||
+		( patternValue instanceof RegExp && patternValue.test( attributeValue ) );
+}
+
+// Checks if attributes of provided element can be matched against provided patterns.
+//
+// @param {Object} patterns Object with information about attributes to match. Each key of the object will be
+// used as attribute name. Value of each key can be a string or regular expression to match against attribute value.
+// @param {module:engine/view/element~Element} element Element which attributes will be tested.
+// @returns {Array|null} Returns array with matched attribute names or `null` if no attributes were matched.
+function matchAttributes( patterns, element ) {
+	// TODO: The `getAttributes()` returns `class` and `style` attribute if present. Should we exclude them when querying attributes?
```

Comment:
I suppose that for now it should return all attributes - it's a bit counterintuitive because we are separating `classes`, `styles` and `attributes` in MatcherPattern, but on the other hand we can request `style` or `class` attribute in old API:
```js
const matcherPattern = { attributes: { style: true, class: true } };
```